groups/ntc/ntco/ntco_ioring.cpp

```diff
@@ -0,0 +1,6359 @@
+            d_memoryMap_p,
+            d_params.submissionQueueOffsetToArray() +
+                d_params.submissionQueueCapacity() * sizeof(bsl::uint32_t));
+        BSLS_ASSERT(rc == 0);
+
+        d_memoryMap_p = 0;
+    }
+}
+
+bsl::uint32_t IoRingSubmissionQueue::head() const
+{
+    LockGuard guard(&d_mutex);
+
+    if (d_memoryMap_p != 0) {
+        NTCO_IORING_READER_BARRIER();
+        return *d_head_p;
+    }
+    else {
+        return 0;
+    }
+}
+
+bsl::uint32_t IoRingSubmissionQueue::tail() const
+{
+    LockGuard guard(&d_mutex);
+
+    if (d_memoryMap_p != 0) {
+        NTCO_IORING_READER_BARRIER();
+        return *d_tail_p;
+    }
+    else {
+        return 0;
+    }
+}
+
+bsl::uint32_t IoRingSubmissionQueue::capacity() const
+{
+    return d_params.submissionQueueCapacity();
+}
+
+IoRingCompletion::IoRingCompletion()
+: d_userData(0)
+, d_result(0)
+, d_flags(0)
+{
+#if NTCO_IORING_COMPLETION_32
+    BSLMF_ASSERT(sizeof(ntco::IoRingCompletion) == 32);
+    d_extraAsQwords[0] = 0;
+    d_extraAsQwords[1] = 0;
+#else
+    BSLMF_ASSERT(sizeof(ntco::IoRingCompletion) == 16);
+#endif
+}
+
+IoRingCompletion::IoRingCompletion(const IoRingCompletion& original)
+: d_userData(original.d_userData)
+, d_result(original.d_result)
+, d_flags(original.d_flags)
+{
+#if NTCO_IORING_COMPLETION_32
+    d_extraAsQwords[0] = original.d_extraAsQwords[0];
+    d_extraAsQwords[1] = original.d_extraAsQwords[1];
+#endif
+}
+
+IoRingCompletion::~IoRingCompletion()
+{
+}
+
+IoRingCompletion& IoRingCompletion::operator=(const IoRingCompletion& other)
+{
+    if (this != &other) {
+        d_userData = other.d_userData;
+        d_result   = other.d_result;
+        d_flags    = other.d_flags;
+#if NTCO_IORING_COMPLETION_32
+        d_extraAsQwords[0] = other.d_extraAsQwords[0];
+        d_extraAsQwords[1] = other.d_extraAsQwords[1];
+#endif
+    }
+
+    return *this;
+}
+
+void IoRingCompletion::reset()
+{
+    d_userData = 0;
+    d_result   = 0;
+    d_flags    = 0;
+#if NTCO_IORING_COMPLETION_32
+    d_extraAsQwords[0] = 0;
+    d_extraAsQwords[1] = 0;
+#endif
+}
+
+ntcs::Event* IoRingCompletion::event() const
+{
+    return reinterpret_cast<ntcs::Event*>(d_userData);
+}
+
+bsl::size_t IoRingCompletion::result() const
+{
+    if (d_result >= 0) {
+        return static_cast<bsl::size_t>(d_result);
+    }
+    else {
+        return 0;
+    }
+}
+
+ntsa::Error IoRingCompletion::error() const
+{
+    if (d_result >= 0) {
+        return ntsa::Error();
+    }
+    else {
+        return ntsa::Error(-d_result);
+    }
+}
+
+bsl::uint8_t IoRingCompletion::flags() const
+{
+    return static_cast<bsl::uint8_t>(d_flags);
+}
+
+bool IoRingCompletion::hasSucceeded() const
+{
+    return d_result >= 0;
+}
+
+bool IoRingCompletion::hasFailed() const
+{
+    return d_result < 0;
+}
+
+bool IoRingCompletion::wasCanceled() const
+{
+    return d_result < 0 && d_result == -ECANCELED;
+}
+
+bool IoRingCompletion::isValid() const
+{
+    return true;
```

Comment:
1) This method seems to be unused
2) Maybe you've wanted to do something like `return d_userData != 0` ?

Reply:
I have removed this function.